### PR TITLE
Added set password flow to browser based SSO

### PR DESCRIPTION
**Pull request by MattTheDev**
## Objective

Added the set password flow into the browser based SSO.

## Code Changes

**src/popup/accounts/set-password.component.html
src/popup/accounts/set-password.component.ts
src/popup/app.module.ts**

New Component

**src/popup/app-routing.module.ts** 

Add routing for new set-password route.

**src/popup/services/services.module.ts**

Crypto Function Service was throwing errors in multiple components, and was coming in as null. Cleaned up the references in this module and it seems to have fixed it..

### CODE REVIEW

File: src/content/sso.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
-window.addEventListener("message", function(event) {
-    if (event.source != window)
+window.addEventListener('message', (event) => {
```
Comment (kspearrin) [inline src/content/sso.ts:1]: I already fixed these lint issues yesterday. You might want to merge master in.

Comment (MattTheDev) [reply src/content/sso.ts]: Done. Just pushed that merge.

File: src/popup/accounts/set-password.component.html [hunk 2]
```diff
@@ -0,0 +1,124 @@
+<form #form (ngSubmit)="submit()" [appApiAction]="formPromise">
+    <header>
+        <div class="left">
+            <a routerLink="/home">{{'cancel' | i18n}}</a>
+        </div>
+        <div class="center">
+            <span class="title">{{'appName' | i18n}}</span>
```
Comment (kspearrin) [inline src/popup/accounts/set-password.component.html:7]: set password. see what web uses.

File: src/popup/accounts/set-password.component.html [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        </div>
+        <div class="right">
+            <button type="submit" appBlurClick [disabled]="form.loading">
+                <span [hidden]="form.loading">{{'login' | i18n}}</span>
```
Comment (kspearrin) [inline src/popup/accounts/set-password.component.html:11]: this isn't logging in, i think just submit. see web

File: src/popup/accounts/set-password.component.html [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            <button type="submit" class="btn primary block" [disabled]="form.loading">
+                <i *ngIf="form.loading" class="fa fa-spinner fa-spin" title="{{'loading' | i18n}}"
+                    aria-hidden="true"></i>
+                <span>{{'submit' | i18n}}</span>
+            </button>
+            <button class="btn block" (click)="logOut()">
+                <span>{{'logOut' | i18n}}</span>
+            </button>
+        </div>
+    </form> -->
```
Comment (kspearrin) [inline src/popup/accounts/set-password.component.html:124]: remove?

File: src/popup/app-routing.module.ts [hunk 5]
```diff
@@ -39,6 +39,7 @@ import { GroupingsComponent } from './vault/groupings.component';
 import { PasswordHistoryComponent } from './vault/password-history.component';
 import { ShareComponent } from './vault/share.component';
 import { ViewComponent } from './vault/view.component';
+import { SetPasswordComponent } from './accounts/set-password.component';
```
Comment (kspearrin) [inline src/popup/app-routing.module.ts:42]: fix order.

File: src/popup/app.module.ts [hunk 6]
```diff
@@ -45,6 +45,7 @@ import { GroupingsComponent } from './vault/groupings.component';
 import { PasswordHistoryComponent } from './vault/password-history.component';
 import { ShareComponent } from './vault/share.component';
 import { ViewComponent } from './vault/view.component';
+import { SetPasswordComponent } from './accounts/set-password.component';
```
Comment (kspearrin) [inline src/popup/app.module.ts:48]: fix order

File: src/popup/app-routing.module.ts (hunk 5 above)
Comment (MattTheDev) [reply src/popup/app-routing.module.ts]: Grouped/Fixed order.

File: src/popup/app.module.ts (hunk 6 above)
Comment (MattTheDev) [reply src/popup/app.module.ts]: Grouped/Fixed order.

File: src/popup/accounts/set-password.component.html (hunk 4 above)
Comment (MattTheDev) [reply src/popup/accounts/set-password.component.html]: Ugh, rookie move. Removed.

File: src/popup/accounts/set-password.component.html (hunk 2 above)
Comment (MattTheDev) [reply src/popup/accounts/set-password.component.html:7]: Matched Web.

File: src/popup/accounts/set-password.component.html (hunk 3 above)
Comment (MattTheDev) [reply src/popup/accounts/set-password.component.html:11]: Matched Web.

File: src/popup/accounts/set-password.component.html [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        </div>
+    </header>
+    <content>
+        <div class="box">
+            <app-callout type="tip">{{'ssoCompleteRegistration' | i18n}}</app-callout>
+            <app-callout type="info" *ngIf="enforcedPolicyOptions">
+                {{'masterPasswordPolicyInEffect' | i18n}}
+                <ul>
+                    <li *ngIf="enforcedPolicyOptions?.minComplexity > 0">
+                        {{'policyInEffectMinComplexity' | i18n : getPasswordScoreAlertDisplay()}}
```
Comment (vincentsalucci) [inline src/popup/accounts/set-password.component.html:23]: This is the first time the browser has dealed with the master password policy enforcement - I believe you'll need to add all of the strings `policyInEffect...`

Comment (MattTheDev) [reply src/popup/accounts/set-password.component.html:23]: Good catch! added.

File: src/popup/accounts/login.component.html [hunk 8]
```diff
@@ -4,11 +4,11 @@
             <a routerLink="/home">{{'cancel' | i18n}}</a>
         </div>
         <div class="center">
-            <span class="title">{{'appName' | i18n}}</span>
+            <span class="title">{{'setMasterPassword' | i18n}}</span>
```
Comment (vincentsalucci) [inline src/popup/accounts/login.component.html:7]: Were these changes intentional? Might've just been done for testing purposes and not reverted.

File: src/popup/accounts/login.component.html [hunk 9]
```diff
@@ -4,11 +4,11 @@
...
         </div>
         <div class="right">
             <button type="submit" appBlurClick [disabled]="form.loading">
-                <span [hidden]="form.loading">{{'login' | i18n}}</span>
+                <span [hidden]="form.loading">{{'submit' | i18n}}</span>
```
Comment (vincentsalucci) [inline src/popup/accounts/login.component.html:11]: Same as above ^

File: src/popup/accounts/login.component.html (hunk 8 above)
Comment (vincentsalucci) [reply src/popup/accounts/login.component.html:7]: I think I see now - you made the changes Kyle requested in this file instead of the `set-password` file.

Comment (MattTheDev) [reply src/popup/accounts/login.component.html:7]: These were mentioned by Kyle in the above feedback. Changed per his request.

File: src/popup/accounts/login.component.html (hunk 9 above)
Comment (MattTheDev) [reply src/popup/accounts/login.component.html:11]: These were mentioned by Kyle in the above feedback. Changed per his request.

File: src/popup/accounts/login.component.html (hunk 8 above)
Comment (MattTheDev) [reply src/popup/accounts/login.component.html:7]: Of course I did.

File: src/popup/accounts/set-password.component.html [hunk 10]
```diff
@@ -0,0 +1,105 @@
+<form #form (ngSubmit)="submit()" [appApiAction]="formPromise">
```
Comment (vincentsalucci) [inline src/popup/accounts/set-password.component.html:1]: Would love a screenshot(s) of this new component in action.

File: src/popup/accounts/login.component.html (hunk 8 above)
Comment (MattTheDev) [reply src/popup/accounts/login.component.html]: Moved to correct file.

File: src/popup/accounts/login.component.html (hunk 9 above)
Comment (MattTheDev) [reply src/popup/accounts/login.component.html]: Moved to correct file.

**Comment by cscharf** [comment]
Since @kspearrin  has requested changes marked on the PR, he'll actually need to make the final approval.

File: src/popup/services/services.module.ts [hunk 11]
```diff
@@ -72,7 +72,7 @@ export const authService = new AuthService(getBgService<CryptoService>('cryptoSe
     messagingService, getBgService<VaultTimeoutService>('vaultTimeoutService')(), null);
 export const searchService = new PopupSearchService(getBgService<SearchService>('searchService')(),
     getBgService<CipherService>('cipherService')(), getBgService<PlatformUtilsService>('platformUtilsService')());
-export const cryptoFunctionService: CryptoFunctionService = new WebCryptoFunctionService(window,
+export const cryptoFunctionService = new WebCryptoFunctionService(window,
     getBgService<PlatformUtilsService>('platformUtilsService')());
```
Comment (kspearrin) [inline src/popup/services/services.module.ts:76]: when was this added? im curious why we are not just referencing the background cryptoFunctionService like before?

Comment (MattTheDev) [reply src/popup/services/services.module.ts]: Duplicate Removed. Using the background service. It wasn't needed.
